gnocchi/rest/api.py

```diff
@@ -901,24 +901,75 @@ def patch(self):
                      if k not in rt_json_current["attributes"]}
         del_attrs = [k for k in rt_json_current["attributes"]
                      if k not in rt_json_next["attributes"]]
+        update_attrs = self.retrieve_update_attrs(rt_json_current,
+                                                  rt_json_next)
 
-        if not add_attrs and not del_attrs:
+        if update_attrs:
+            LOG.debug("Updating attributes [%s] for resource-type [%s]",
+                      update_attrs, self._name)
+
+        if not add_attrs and not del_attrs and not update_attrs:
             # NOTE(sileht): just returns the resource, the asked changes
             # just do nothing
             return rt
 
         try:
             add_attrs = schema.attributes_from_dict(add_attrs)
+            update_attrs = self.create_update_attrs(schema, update_attrs)
         except resource_type.InvalidResourceAttribute as e:
             abort(400, "Invalid input: %s" % e)
 
         try:
             return pecan.request.indexer.update_resource_type(
                 self._name, add_attributes=add_attrs,
-                del_attributes=del_attrs)
+                del_attributes=del_attrs, update_attributes=update_attrs)
         except indexer.NoSuchResourceType as e:
             abort(400, six.text_type(e))
 
+    def create_update_attrs(self, schema, update_attrs):
+        try:
+            new_attrs = dict(map(lambda entry: (entry[0], entry[1][1]),
+                                 update_attrs.items()))
+            old_attrs = dict(map(lambda entry: (entry[0], entry[1][0]),
+                                 update_attrs.items()))
+            update_attrs_new = schema.attributes_from_dict(new_attrs)
+            update_attrs_new.sort(key=lambda attr: attr.name)
+            update_attrs_old = schema.attributes_from_dict(old_attrs)
+            update_attrs_old.sort(key=lambda attr: attr.name)
+            update_attrs = []
+            for i in range(len(update_attrs_new)):
+                update_attrs.append((update_attrs_new[i],
+                                     update_attrs_old[i]))
+
+            return update_attrs
+        except Exception:
+            abort(400, "Unable to update attributes [%s]." % update_attrs)
```

Comment:
That does not look like a 400: if it's a 400 you must be able to explain what's wrong to the client.
If you're unable to execute the request it's more likely a 500 and a bug in the code itself.
I'd rather not try/except anything here.